File: internal/models/portscan.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// PortScanStatus 定义端口扫描状态
type PortScanStatus string

const (
	// PortScanStatusPending 等待扫描
	PortScanStatusPending PortScanStatus = "pending"
	// PortScanStatusRunning 扫描中
	PortScanStatusRunning PortScanStatus = "running"
	// PortScanStatusCompleted 扫描完成
	PortScanStatusCompleted PortScanStatus = "completed"
	// PortScanStatusFailed 扫描失败
	PortScanStatusFailed PortScanStatus = "failed"
)

// PortScanTask 定义端口扫描任务
type PortScanTask struct {
	ID            primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	ProjectID     primitive.ObjectID `bson:"projectId" json:"projectId"`
	Name          string             `bson:"name" json:"name"`
	Description   string             `bson:"description" json:"description"`
	Targets       []string           `bson:"targets" json:"targets"`
	Ports         []int              `bson:"ports" json:"ports"`
	Protocol      string             `bson:"protocol" json:"protocol"` // TCP, UDP
	RateLimit     int                `bson:"rateLimit" json:"rateLimit"`
	Timeout       int                `bson:"timeout" json:"timeout"`
	Status        PortScanStatus     `bson:"status" json:"status"`
	Progress      float64            `bson:"progress" json:"progress"`
	StartTime     time.Time          `bson:"startTime" json:"startTime"`
	EndTime       time.Time          `bson:"endTime" json:"endTime"`
	CompletedAt   time.Time          `bson:"completedAt" json:"completedAt"`
	CreatedAt     time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt     time.Time          `bson:"updatedAt" json:"updatedAt"`
	CreatedBy     string             `bson:"createdBy" json:"createdBy"`
	NodeID        string             `bson:"nodeId" json:"nodeId"`
	Error         string             `bson:"error" json:"error"`
	ResultSummary PortScanSummary    `bson:"resultSummary" json:"resultSummary"`
	Config        PortScanConfig     `bson:"config" json:"config"`
}

// PortScanResult 定义端口扫描结果
type PortScanResult struct {
	ID        primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	TaskID    primitive.ObjectID `bson:"taskId" json:"taskId"`
	ProjectID primitive.ObjectID `bson:"projectId" json:"projectId"`
	Host      string             `bson:"host" json:"host"`
	IP        string             `bson:"ip" json:"ip"`
	Port      int                `bson:"port" json:"port"`
	Protocol  string             `bson:"protocol" json:"protocol"` // TCP, UDP
	Status    string             `bson:"status" json:"status"`     // open, closed, filtered
	Service   string             `bson:"service" json:"service"`
	Product   string             `bson:"product" json:"product"`
	Banner    string             `bson:"banner" json:"banner"`
	Version   string             `bson:"version" json:"version"`
	ExtraInfo string             `bson:"extraInfo" json:"extraInfo"`
	TTL       int                `bson:"ttl" json:"ttl"`
	Reason    string             `bson:"reason" json:"reason"`
	CreatedAt time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt time.Time          `bson:"updatedAt" json:"updatedAt"`
}

// PortScanConfig 定义端口扫描配置
type PortScanConfig struct {
	DefaultPorts     []int    `bson:"defaultPorts" json:"defaultPorts"`
	DefaultProtocol  string   `bson:"defaultProtocol" json:"defaultProtocol"`
	DefaultRateLimit int      `bson:"defaultRateLimit" json:"defaultRateLimit"`
	DefaultTimeout   int      `bson:"defaultTimeout" json:"defaultTimeout"`
	MaxTargets       int      `bson:"maxTargets" json:"maxTargets"`
	MaxPorts         int      `bson:"maxPorts" json:"maxPorts"`
	EnableBanner     bool     `bson:"enableBanner" json:"enableBanner"`
	EnableService    bool     `bson:"enableService" json:"enableService"`
	EnableVersion    bool     `bson:"enableVersion" json:"enableVersion"`
	RateLimit        int      `bson:"rateLimit" json:"rateLimit"`
	Ports            string   `bson:"ports" json:"ports"`
	ScanType         string   `bson:"scanType" json:"scanType"`
	Concurrency      int      `bson:"concurrency" json:"concurrency"`
	Timeout          int      `bson:"timeout" json:"timeout"`
	ScanMethod       string   `bson:"scanMethod" json:"scanMethod"`
	ServiceDetection bool     `bson:"serviceDetection" json:"serviceDetection"`
	ExcludeHosts     []string `bson:"excludeHosts" json:"excludeHosts"`
	SaveToDB         bool     `bson:"saveToDB" json:"saveToDB"`
	RetryCount       int      `bson:"retryCount" json:"retryCount"`
}

// PortScanSummary 定义端口扫描摘要
type PortScanSummary struct {
	TotalHosts     int            `bson:"totalHosts" json:"totalHosts"`
	ProcessedHosts int            `bson:"processedHosts" json:"processedHosts"`
	OpenPorts      int            `bson:"openPorts" json:"openPorts"`
	ServiceStats   map[string]int `bson:"serviceStats" json:"serviceStats"`
}

File: web/src/lib/types/portscan.ts
/**
 * 端口扫描相关类型定义
 */

/**
 * 端口扫描任务配置
 */
export interface PortScanConfig {
	target: string; // 扫描目标（IP或域名）
	ports: string; // 端口配置（如 "80,443,1-1000"）
	scanMethod?: 'tcp' | 'udp' | 'both'; // 扫描方法
	maxWorkers?: number; // 最大并发数
	timeout?: number; // 超时时间（秒）
	enableBanner?: boolean; // 启用banner抓取
	enableSSL?: boolean; // 启用SSL检测
	enableService?: boolean; // 启用服务识别
	rateLimit?: number; // 速率限制（每秒请求数）
}

/**
 * 端口扫描任务
 */
export interface PortScanTask {
	id: string;
	name: string;
	description?: string;
	type: 'port_scan';
	status: 'pending' | 'queued' | 'running' | 'completed' | 'failed' | 'canceled';
	priority: number;
	config: PortScanConfig;
	projectId?: string;
	nodeId?: string;
	createdBy: string;
	createdAt: string;
	updatedAt: string;
	startTime?: string;
	endTime?: string;
	progress: TaskProgress;
	result?: TaskResult;
	error?: string;
}

/**
 * 任务进度
 */
export interface TaskProgress {
	total: number;
	current: number;
	completed: number;
	failed: number;
	percentage: number;
	speed?: number; // 每秒处理数量
	estimatedTime?: number; // 预计剩余时间（秒）
}

/**
 * 任务结果
 */
export interface TaskResult {
	openPorts: PortScanResult[];
	scannedPorts: PortScanResult[];
	openCount: number;
	totalScanned: number;
	target: string;
	scanMethod: string;
	serviceStats: Record<string, number>;
	scanDuration: number;
	summary: string;
}

/**
 * 端口扫描结果
 */
export interface PortScanResult {
	host: string;
	port: number;
	status: 'open' | 'closed' | 'filtered';
	protocol: 'tcp' | 'udp';
	service?: string;
	banner?: string;
	version?: string;
	responseTime: number;
	timestamp: string;
	error?: string;
	sslInfo?: SSLInfo;
	fingerprint?: Record<string, string>;
}

/**
 * SSL证书信息
 */
export interface SSLInfo {
	issuer: string;
	subject: string;
	notBefore: string;
	notAfter: string;
	fingerprint: string;
	version: string;
	cipher: string;
}

/**
 * 任务创建请求
 */
export interface TaskCreateRequest {
	name: string;
	description?: string;
	target: string;
	ports: string;
	scanMethod?: 'tcp' | 'udp' | 'both';
	maxWorkers?: number;
	timeout?: number;
	enableBanner?: boolean;
	enableSSL?: boolean;
	enableService?: boolean;
	rateLimit?: number;
	projectId?: string;
}

/**
 * 任务列表响应
 */
export interface TaskListResponse {
	tasks: PortScanTask[];
	total: number;
	page: number;
	limit: number;
	totalPages: number;
}

/**
 * 端口扫描表单数据
 */
export interface PortScanFormData {
	name: string;
	description: string;
	target: string;
	ports: string;
	preset: string;
	scanMethod: 'tcp' | 'udp' | 'both';
	advanced: {
		maxWorkers: number;
		timeout: number;
		enableBanner: boolean;
		enableSSL: boolean;
		enableService: boolean;
		rateLimit: number;
	};
	projectId?: string;
}

/**
 * 端口扫描统计数据
 */
export interface PortScanStatistics {
	totalTasks: number;
	completedTasks: number;
	failedTasks: number;
	totalPorts: number;
	openPorts: number;
	successRate: number;
	averageDuration: number;
	serviceStats: Record<string, number>;
	recentTasks: PortScanTask[];
	topTargets: Array<{
		target: string;
		count: number;
		lastScan: string;
	}>;
}

/**
 * 端口预设配置
 */
export interface PortPreset {
	id: string;
	name: string;
	ports: string;
	description: string;
	category: 'common' | 'web' | 'database' | 'service' | 'security';
}

/**
 * 扫描任务过滤器
 */
export interface TaskFilter {
	status?: string;
	target?: string;
	projectId?: string;
	dateRange?: {
		startDate: string;
		endDate: string;
	};
	createdBy?: string;
}

/**
 * 端口扫描配置验证结果
 */
export interface ValidationResult {
	valid: boolean;
	errors: string[];
	warnings: string[];
	estimatedDuration?: number;
	portCount?: number;
}

/**
 * 实时进度更新事件
 */
export interface ProgressUpdateEvent {
	taskId: string;
	progress: TaskProgress;
	latestResults?: PortScanResult[];
	status: string;
}

/**
 * 导出配置
 */
export interface ExportConfig {
	format: 'csv' | 'json' | 'xlsx' | 'xml';
	includeDetails: boolean;
	onlyOpenPorts: boolean;
	customFields?: string[];
}

/**
 * 服务指纹信息
 */
export interface ServiceFingerprint {
	service: string;
	version?: string;
	product?: string;
	extraInfo?: string;
	confidence: number;
	cpe?: string;
}
